src/spaces/jolt_debug_geometry_3d.cpp

```diff
@@ -115,8 +123,44 @@ JoltDebugGeometry3D::~JoltDebugGeometry3D() = default;
 
 #endif // JPH_DEBUG_RENDERER
 
+JPH::RefConst<JPH::Shape> JoltDebugGeometry3D::remove_custom_shapes(const JPH::Shape* p_shape) {
+	switch (p_shape->GetSubType()) {
+		case JoltCustomShapeSubType::EMPTY:
+		case JoltCustomShapeSubType::RAY:
+		case JoltCustomShapeSubType::MOTION: {
+			// Replace unsupported shapes with a small sphere
+			return new JPH::SphereShape(0.1f);
+		}
+
+		case JoltCustomShapeSubType::OVERRIDE_USER_DATA:
+		case JoltCustomShapeSubType::DOUBLE_SIDED: {
+			// Replace unsupported decorator shapes with the inner shape
+			return static_cast<const JPH::DecoratedShape*>(p_shape)->GetInnerShape();
+		}
+
+		case JPH::EShapeSubType::Scaled: {
+			const JPH::ScaledShape* scaled_shape = static_cast<const JPH::ScaledShape*>(p_shape);
+			JPH::RefConst<JPH::Shape> new_shape = remove_custom_shapes(scaled_shape->GetInnerShape());
+			if (scaled_shape->GetInnerShape() != new_shape) {
+				return new JPH::ScaledShape(new_shape, scaled_shape->GetScale());
+			}
+			return p_shape;
+		}
+
+		case JPH::EShapeSubType::StaticCompound:
+		case JPH::EShapeSubType::MutableCompound:
+		case JPH::EShapeSubType::RotatedTranslated:
+		case JPH::EShapeSubType::OffsetCenterOfMass: {
+			ERR_PRINT("Unsupported shape type, snapshot may not be readable in Jolt!");
+			return p_shape;
+		}
```

Comment:
What's with needing to special-case the decorator and compound shapes by the way? Is `SaveBinaryState` not able to deal with those types of shapes?

Both `RotatedTranslated` and `StaticCompound` will be quite prevalent. `OffsetCenterOfMass` not so much perhaps, but it's used as well.

Reply:
Yes SaveBinaryState would work but the Jolt Samples app doesn't know about to these shapes so will fail loading. As most shapes are immutable I need to create new shapes, and I only implemented the ones in the scene I had to export (not knowing if you would like to merge the code or not, if you do I can create code for the other cases too)

Reply:
Oh right, you need to create new shapes because they could be wrapping/containing any of the custom shapes. That makes sense.

I'll see if I can implement them while I'm at it.